---
title: "FILLINStartup"
author: "Sarah Odell"
date: "5/14/2018"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,eval=FALSE)
```
#Imputation Using FILLIN

Getting FILLIN up and running, and using the maize NAM founders to impute onto the NAM RILs
which were genotyped using GBS (v2.7)

User Manual on FILLIN Imputation tool: https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/FILLIN/FILLIN

###Getting Started
Download TASSEL from: https://bitbucket.org/tasseladmin/tassel-5-source/wiki/Home
(requires Java 8)
```{bash gettassel}
git clone https://bitbucket.org/tasseladmin/tassel-5-standalone.git

#I recommend doing
#PATH=$PATH:[whatever you downloaded tassel]
```
Other Software Requirements:
Java verion 1.8
icommands/1
samtools/1.3.1
bwa/0.7.9a
tabix/0.2.6
picardtools/2.7.1
bcftools/1.2

Files available for download from Cyverse using Cyberduck or icommands (must make a Cyverse account)
```{bash icommand}
#use iinit command to start/set up an icommands account
#GBS v2.7
iget -K -P -T /iplant/home/shared/panzea/genotypes/GBS/v27/ZeaGBSv27_publicSamples_rawGenos_AGPv3_20170206.h5
#HapMap v3.21 Chromosome 10 (Unimputed, aligned to AGPv3)
iget -K -P -T /iplant/home/shared/panzea/hapmap3/hmp321/unimputed/c10_hmp321_withGL.vcf.gz
```

###Filter GBS data
Convert the GBS hdf5 file to hmp, and filter using the taxa list file 'NAM_RILS.json.gz' so we have only the ~5,000 RILS. We also use the HetsToUnknown Plugin to remove heterozygous calls. This will make imputation easier.

```{bash h5tovcf}
run_pipeline.pl -Xmx16g -h5 ZeaGBSv27_publicSamples_rawGenos_AGPv3_20170206.h5 -FilterTaxaBuilderPlugin -t
taxaList NAM_RILS.json.gz -endPlugin -HetsToUnknown -endPlugin -export NAMRILS_ZeaGBSv27_APGv3_noHets.hmp.txt.gz -exportType Hapmap
```

Then, get only chromosome 10. Here's just a simple script with python. There may be a more efficient way to filter out chromosomes with Tassel or 
some other software, but this works \_('.')_/
```{python get10, python.reticulate=FALSE }
import gzip

with gzip.open('NAMRILS_ZeaGBSv27_APGv3_noHets.hmp.txt.gz','r') as infile:
        txt=''
        for line in infile:
            if 'rs#' in line:
                txt+=line
            else:
                cols = line.split('\t')
                if cols[2] == 10:
                    txt+=line
    with gzip.open('NAMRILS_ZeaGBSv27_chr10.hmp.txt.gz','w') as outfile:
        outfile.write(txt)
```

### Filter HapMap files

The HapMap files are missing a line that allows them to be indexed with tabix.
It should look like this:
"##contig=<ID=10,length=149632204>"
I used the B73 APGv3 (available here: ftp://ftp.ensemblgenomes.org/pub/release-31/plants/fasta/zea_mays/dna/)
to get the contig length for chromosome 10. 
```{bash indexv3}
#Index the reference genome chromosome 10
gunzip -k Zea_mays.AGPv3.31.dna.chromosome.10.fa.gz
bwa index -a bwtsw Zea_mays.AGPv3.31.dna.chromosome.10.fa

samtools faidx Zea_mays.AGPv3.31.dna.chromosome.10.fa

java -jar picard.jar CreateSequenceDictionary REFERENCE=Zea_mays.AGPv3.31.dna.chromosome.10.fa OUTPUT=Zea_mays.AGPv3.31.dna.chromosome.10.dict


awk '{printf("##contig=<ID=%s,length=%d>\\n",$1,$2);}' Zea_mays.AGPv3.31.dna.chromosome.10.fa.fai > chrom_length.txt
```

The output of this should be 
"##contig=<ID=10,length=149632204>"
This awk command will add this line to our hapmap file

```{bash hmpv3}
gunzip -k c10_hmp321_withGL.vcf.gz
awk '/^#CHROM/ {printf("##contig=<ID=10,length=149632204>\n");} {print;}' c10_hmp321_withGL.vcf > c10_hmp31_edit.vcf
```

Getting only the NAM founders from the HapMap file
The file nam_founders.txt just contains the names of the NAM founders. bcftools can be used to get
only those samples from the vcf file
```{bash getfounders}
#zip and index the vcf file
bgzip c10_hmp31_edit.vcf 
tabix -p vcf c10_hmp31_edit.vcf.gz
#extract the founders
bcftools view -Oz -S nam_founders.txt c10_hmp31_edit.vcf.gz > c10_hmp31_edit_founders.vcf.gz
```

### Build FILLIN Haplotypes

The founders HapMap data can then be used to build haplotypes using the FILLINFindHaplotypesPlugin
Information on option can be found in the TASSEL User Manual (link at top). -minTaxa 1 is used instead of the default 2 because we want haplotypes to be build for individual founders, if possible. -mxErr 0 because we don't want founders to be lumped into the same haplotype if there are differences between them.
This will take a while (~9 hours)
```{bash buildhaplos}
run_pipeline.pl -Xmx32g -FILLINFindHaplotypesPlugin -hmp c10_hmp31_edit_founders.vcf.gz -o NAMHaps/ -mxErr 0 -minTaxa 1 -extOut true -endPlugin
```

### Impute NAM RILs 
```{bash impute}
run_pipeline.pl -Xmx32g -FILLINImputationPlugin -hmp NAMRILS_ZeaGBSv27_APGv3_noHets.hmp.txt.gz -d NAMHaps/ -o NAMRILS_ZeaGBSv27_APGv3_FILLINImputed.hmp.txt.gz -endPlugin
```

To get information on which founder was assigned to each haplotype break, use the -ProjA flag.
This outputs a projection alignment file, showing where chromosome breaks were designated and which
donor was imputed where.
The donor (-d) must then be the original donor file that was used to generate haplotypes (high SNP density)
```{bash projA}
run_pipeline.pl -Xmx32g -FILLINImputationPlugin -hmp NAMRILS_ZeaGBSv27_APGv3_noHets.hmp.txt.gz -d c10_hmp31_edit_founders.vcf.gz -o NAMRILS_ZeaGBSv27_APGv3_FILLINImputed_ProjA.hmp.txt.gz -ProjA true -endPlugin
```

