---
title: "IBD_Corr_Filter"
author: "Sarah Odell"
date: "8/17/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Filtering Markers based on IBD Segment and Correlation of Genotype Probabilities
Using Chromosome 10 for 10 DH lines

```{r setup}
library('qtl2')
library("tidyverse")
library("dplyr")
library("data.table")
library("ggplot2")
library("reshape2")
library("tibble")
library('igraph')
library('abind')
setwd("~/Documents/PBGG/MAGIC/impute")
```

Read in files for genotype probabilities and IBD segments

``` {r files}
pr10=readRDS('../qtl2/Biogemma_071118/bg10_geno_probs.rds')
ibd=readRDS('../qtl2/Biogemma_071118/bg10_ibdsegments.rds')
```

Match letter codes to founder names for easier comparison
``` {r founders}
founders=c("A632_usa","B73_inra","CO255_inra","FV252_inra","OH43_inra","A654_inra","FV2_inra","C103_inra","EP1_inra","D105_inra","W117_inra","B96","DK63","F492","ND245","VA85")
founderfile=fread('../qtl2/Founder_colorcodes.txt',header=F)
names(founderfile)=c('line','hex')
code=c("A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P")
founderfile$code=code
i1 <- match(ibd$strain1,founderfile$code)
i2 <- match(ibd$strain2,founderfile$code)
ibd$line1=founderfile$line[i1]
ibd$line2=founderfile$line[i2]

ibd$pair=do.call(paste,c(ibd[c('line1','line2')],sep="-"))
```


We want to identify unique haplotype regions where a set number of founders are in IBD with one another. This will allow us to collapse the number of haplotypes in this regions from 16 to however many unique haplotypes are within that region (i.e. if 4 founders are in IBD in this region, then there are 13 unique haplotypes).

Build and adjacency matrix of IBD pairs and
Count number of 'islands' from the adjacency matrix.
Unique haplotype groups are assigned to a particular integer

```{r ibd_segments}
ibd=ibd[order(ibd$left_pos),]
rownames(ibd)=seq(1,nrow(ibd))

ibd_segments=c()
start=min(ibd$left_pos) #start with first segment in chromosome
ibd_graph=array(0,dim=c(16,16,1))
while(start<max(ibd$right_pos)){
  remaining=ibd[(ibd$left_pos>start | ibd$right_pos>=start),] #grab segments that are to the right of the start
  remaining$left_dist=remaining$left_pos - start 
  remaining$right_dist=remaining$right_pos - start
  lowest_left=min(remaining$left_dist[remaining$left_dist>0])
  lowest_right=min(remaining$right_dist[remaining$right_dist>0])
  closest=min(lowest_left,lowest_right) #find the next start or end of segment that is closest to current start
   if (lowest_left==closest){
    next_break=remaining[remaining$left_dist==closest,]$left_pos[1]
  }
  if (lowest_right==closest){
    next_break=remaining[remaining$right_dist==closest,]$right_pos[1]
  }
  within=remaining[(remaining$right_pos>=next_break) & (remaining$left_pos<=start),] #identify ibd segments that are between start and next_break

  if(dim(within)[1]>=1){
    adj=matrix(0,nrow=16,ncol=16,dimnames=list(founders,founders))
    for (i in 1:nrow(within)){
      adj[c(within[i,'line1']),c(within[i,'line2'])]=1
      adj[c(within[i,'line2']),c(within[i,'line1'])]=1
    }
    ibd_graph=abind(ibd_graph,adj)
    graph=graph_from_adjacency_matrix(adj,mode=c('undirected'))
    blocks=unname(components(graph)$membership)
    n_grps=components(graph)$no
  }
  else{
    blocks=seq(1,16)
    n_grps=16
  }
  ibd_segments=rbind(ibd_segments,c(10,start,next_break,blocks,n_grps))
  start=next_break
}
ibd_segments=as.data.frame(ibd_segments)
names(ibd_segments)=c('chrom','start','end',founders,'n_haps')
```
I wanted to select the haplotype segments which were "incomplete" in terms of IBD pairs (A and B, B and C are in IBD, but not A and C). If a haplotype group is complete, the number of edges in the graph connecting them should be n choose 2, where n is the number of lines in the haplotype group. 
I compared the haplotype block size for the incomplete and complete blocks. They do not seem to be much different in size. There are 1929 such incomplete segments in chromosome 10. I do not know whether or not this will be a problem.

```{r not_complete}
not_complete=c()
for(i in seq(1,dim(ibd_graph)[3])){
  adj=ibd_graph[,,i]
  graph=graph_from_adjacency_matrix(adj,mode=c('undirected'))
  n_grps=components(graph)$no
  for(j in seq(1,n_grps)){
      size=components(graph)$csize[j]
      if(size>2){
        grp=graph_from_adjacency_matrix(adj[components(graph)$membership==j,components(graph)$membership==j],mode=c('undirected'))
        if(choose(size,2)>gsize(grp)){
          not_complete=c(not_complete,i)
        }
      }
  }
}

incomplete=seq(1,dim(ibd_segments)[1]) %in% not_complete
ibd_segments$incomplete=incomplete
ibd_segments$segsize=ibd_segments$end-ibd_segments$start

sub=ibd_segments[ibd_segments$segsize<1e6,]

ggplot(sub,aes(x=incomplete,y=segsize)) + geom_boxplot(aes(color=incomplete))
```




Write out this table to a file for later use

``` {r write}
fwrite(ibd_segments,file="DH10_Chr10_IBD_Regions.txt",sep='\t',row.names = F,quote=F)
```

Lets create a plot to show how the number of unique haplotypes 
changes as we move along the chromosome.
```{r visualize}
hex_colors=c("#f42896","#84ef7c","#a8bc44","#8ed1d6","#702349",
             "#f2875b","#28ad26","#afd3ef","#937266","#56cc59",
             "#663dd3","#478959","#47145b","#7c2126","#ad147a",
             "#afb735")

#png('Chr10_haplotypes.png',width = 920, height = 480)
print(ggplot(ibd_segments, aes(start, n_haps)) +
  geom_segment(aes(xstart=start,xend = end,color="red", ystart=n_haps,yend=n_haps),lineend="butt",size=10) +
  ggtitle("Number of Unique Chromosome 10 Haplotypes") + xlab("Position (Mb)") + ylab("Haplotype Number")+guides(color=F))
#dev.off()

```


As you can see, there is a lot of variation in haplotype number. There are some regions, particularly around the centromere, where the number of unique haplotypes in the founders is very low.

``` {r format}
ibd_segments=fread("../qtl2/Biogemma_071118/DH10_Chr10_IBD_Regions.txt",data.table=F)
pmap=fread('../qtl2/Biogemma_071118/startfiles/Biogemma_pmap_c10.csv',data.table=F)
pmap$pos=pmap$pos*1e6
dimnames(pr10[[1]])[[2]]=founders
samples=unlist(dimnames(pr10[[1]])[1])

final_haplo=list(chr=10)

groups=sort(unique(ibd_segments$n_haps))
for(h in groups){
  n_hap=h
  n_ind=dim(pr10[[1]])[1]
  # number of IBD Segments with 15 unique haplotypes
  test=ibd_segments[ibd_segments$n_haps==n_hap,]
  rownames(test)=seq(1,nrow(test))
  # for each individual
  for(n in seq(1,n_ind)){
    hprobs=c()
    markers=c()
    #go through each haplotype block with n_hap haplotype groups
    for(i in seq(1,nrow(test))){
      start=test$start[i]
      end=test$end[i]
      # grab the SNPs within these segments
      within=pmap[(pmap$pos < end) & (pmap$pos >= start),]
      if(dim(within)[1]!=0){
        # get the founder probabilities of these SNPs
        prob=pr10[[1]][n,,within$marker]
        line=test[i,]
        markers=c(markers,within$marker)
        # for each haplotype group grab the founders in that haplotype group
        # and sum the probability
        #if(length(dim(prob))==2){
        hprob=sapply(seq(1,n_hap), function(x) t(prob) %*% as.vector(line[,founders]==x))
        #else{
        #  hap_prob=sapply(seq(1,n_hap), function(x) t(prob[n,,]) %*% as.vector(line[,founders]==x))
        #}
        hprobs=rbind(hprobs,hprob)
      }
    }
    a=dim(hprobs)[1]
    b=dim(hprobs)[2]
    hprob_array=array(hprobs,dim=c(1,a,b))
    if(n==1){
      haplo_probs=array(hprob_array,dim=c(n_ind,a,b))
    }
    else{
      haplo_probs[n,,]=hprob_array
    }
  }
  dimnames(haplo_probs)=list(samples,markers,seq(1,h))
  final_haplo[[h]]=haplo_probs
  
}

# for each IBD segment

saveRDS(final_haplo,'Haplotype_probablities_chr10.rds')

```



```{r blah}
haplo=readRDS('Haplotype_probablities_chr10.rds')

nine=haplo[[9]]
dim(nine)

all_pr=c()
for (m in 1:10){
  td <- nine[m,,]
  mdf <- as.data.frame(td,row.names = rownames(td))
  mdf<-rownames_to_column(mdf,"marker")
  #names(mdf)=c("marker",founders)
  mdf <- merge(mdf,pmap,by.x='marker',by.y='marker')
  mdf <- mdf[,c(2:10,12)]
  mlong<-melt(mdf,id="pos")
  mlong$sample=rep(samples[m],dim(mlong)[1])
  all_pr=rbind(all_pr,mlong)
}

hex_colors=c("#f42896","#84ef7c","#a8bc44","#8ed1d6","#702349",
             "#f2875b","#28ad26","#afd3ef","#937266","#56cc59",
             "#663dd3","#478959","#47145b","#7c2126","#ad147a",
             "#afb735")
colors=hex_colors[1:9]

conf=all_pr[all_pr$value>=0.95,]

png('HapGrps9_Chr10.png',width=920,height=640)
ggplot(data=all_pr,aes(x=pos,y=value,color=variable)) + facet_grid(sample ~ .) +scale_color_manual(values=colors) + scale_fill_manual(values=colors)+ geom_ribbon(aes(ymin=0,ymax=value,fill=variable),alpha=5/10) + geom_line() + ggtitle("Haplotype Probability for 10 DH Lines") + xlab("Position (Mb)") + ylab("Probability") + scale_y_continuous(breaks=c(0.0,0.5,1.0)) + guides(color=FALSE)
dev.off()
```